File: Blum Blum Shub/bbs.py
def is_prime(n):
    if n <= 1: return False
    if n <= 3: return True
    if n % 2 == 0 or n % 3 == 0: return False
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0: return False
        i += 6
    return True

def blum_blum_shub(p, q, seed, n_bits):
    assert is_prime(p) and p % 4 == 3, "p must be prime and ≡ 3 mod 4"
    assert is_prime(q) and q % 4 == 3, "q must be prime and ≡ 3 mod 4"
    M = p * q
    assert 1 < seed < M and seed % p != 0 and seed % q != 0, "Invalid seed"

    x = seed
    bits = []

    for _ in range(n_bits):
        x = pow(x, 2, M)
        bits.append(x % 2)

    return bits

# Example usage
p = 499
q = 547
seed = 159201
n = 20

output_bits = blum_blum_shub(p, q, seed, n)
print("Random bits:", ''.join(map(str, output_bits)))

File: Blum Blum Shub/bbs.cpp
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

bool isPrime(int n) {
    if (n <= 1) return false;
    if (n <= 3) return true;
    if (n % 2 == 0 || n % 3 == 0) return false;
    for (int i = 5; i * i <= n; i += 6)
        if (n % i == 0 || n % (i + 2) == 0) return false;
    return true;
}

vector<int> blumBlumShub(int p, int q, int seed, int n_bits) {
    assert(isPrime(p) && p % 4 == 3);
    assert(isPrime(q) && q % 4 == 3);

    long long M = 1LL * p * q;
    assert(seed > 1 && seed < M && seed % p != 0 && seed % q != 0);

    long long x = seed;
    vector<int> bits;

    for (int i = 0; i < n_bits; ++i) {
        x = (x * x) % M;
        bits.push_back(x % 2);
    }

    return bits;
}

int main() {
    int p = 499, q = 547;
    int seed = 159201;
    int n = 20;

    vector<int> bits = blumBlumShub(p, q, seed, n);
    cout << "Random bits: ";
    for (int b : bits) cout << b;
    cout << endl;

    return 0;
}
